### PR TITLE
[Core] Explicitly manage torch CPU threads in workers

### LINKED ISSUE #49013: [Perf] ~2x decode throughput regression for structured outputs since #45424: apply_grammar_bitmask staging rewrite (bisected to commit, file, and hunk)
## Summary

End-to-end decode throughput for guided-JSON (xgrammar) workloads drops ~2x
between v0.23.0 and v0.24.0 (also v0.25.1). Bisected to commit
**7df3d7dada840c68b85b26b79de7f59f676d58e3** "[Core] Ensure memory is pinned
prior to async h2d copy" (#45424), then within the commit to a single file:
**vllm/v1/structured_output/utils.py — `apply_grammar_bitmask`**.

Evidence chain (identical harness, deterministic to ±1s per 2,000-request
drain; A100 80GB, Gemma-4-26B-A4B-IT bf16, max_num_seqs=64, ngram spec decode
k=5, CUDA graphs, EVERY request uses output_format json_schema → xgrammar):

| build | drain time | rate |
|---|---|---|
| 8dd1b702f (parent commit) | 513 s | 233 req/min |
| 7df3d7dad (culprit) | 985 s | 121 req/min |
| 7df3d7dad + ALL 29 v1/* files reverted to parent | 534 s | 224 req/min |
| 7df3d7dad + ONLY structured_output/utils.py reverted | 534 s | 224 req/min |
| 7df3d7dad + worker files (gpu_model_runner/gpu_input_batch) reverted | 1006 s | 119 req/min |
| 7df3d7dad + PIN_MEMORY forced False (torch_utils) | 985 s | 121 req/min |
| v0.24.0 release | 7581 s vs 3810 s on 15k corpus | 2.0x slower |

Note the PIN_MEMORY=False row: forcing the new global pin flag off does NOT
recover throughput, so the cost is in the staging rewrite itself, not merely
the pinned allocation flag.
Hunk-level split: restoring ONLY the staging block below (keeping the
commit's other change in the same function — index_tensor via
async_tensor_h2d) fully recovers baseline: **514 s / 233 req/min**. The
staging rewrite is the entire regression.

## The suspect code

#45424 changed the per-step bitmask staging in `apply_grammar_bitmask` from:

```python
sorted_bitmask = np.full(
    shape=(logits.shape[0], grammar_bitmask.shape[1]), fill_value=-1,
    dtype=grammar_bitmask.dtype,
)
...  # reorder loop fills rows
grammar_bitmask = torch.from_numpy(sorted_bitmask).to(logits.device, non_blocking=True)
```

to:

```python
sorted_bitmask_tensor = torch.full(
    (logits.shape[0], grammar_bitmask.shape[1]), -1,
    dtype=torch.from_numpy(grammar_bitmask[:0]).dtype,
    pin_memory=PIN_MEMORY,
)
sorted_bitmask = sorted_bitmask_tensor.numpy()
...  # same reorder loop, now writing into the (pinned) tensor's numpy view
grammar_bitmask = sorted_bitmask_tensor.to(logits.device, non_blocking=True)
```

For a 262k-vocab model at max_num_seqs=64 with k=5 ngram speculation this is a
~(384 x 8192) int32 = ~12.6 MB freshly allocated (normally pinned) host tensor
per engine step, plus the reorder loop writing through the numpy view of that
pinned buffer, freed again at the end of the step. This path runs every step
for every batch containing structured-output requests.

## Setup detail

- A100 80GB PCIe, torch 2.11.0 (cu13), python 3.12, offline LLM engine
  (in-process), xgrammar backend, chunked prefill 8192, TRITON_ATTN
  (forced by Gemma4 hetero head dims), async scheduling disabled (ngram).
- Symptom profile: engine steps uniformly ~2x slower at identical batch
  occupancy, acceptance rates, KV/prefix-cache hit, GPU clocks; CUDA graphs
  on/off preserves the gap; regression also present WITHOUT spec decode
  (0.25.1 no-ngram run degraded >2x as well). All requests guided-JSON.
- Release-level numbers (14,985-request production-derived corpus):
  v0.21.0 3810 s makespan / 531 accepted tok/s → v0.24.0 7581 s / 266,
  v0.25.1 7590 s / 265.

## Suggested fix

Restore the numpy staging (or keep a cached, reused pinned staging buffer
sized to (max_logits_rows, bitmask_width) instead of a fresh per-step
allocation).

Possibly related: #47047 (Gemma-4-26B-A4B on SM120/NVFP4, 0.21→0.23 slowdown
at low concurrency, unresolved — different window but same model family and
also structured-output-heavy workloads).

Full bisect logs (9 commit-level probes via per-commit wheels + 8 file/hunk-level
image probes) available on request.

### PR DESCRIPTION
**Pull request by njhill**
Clamp for startup, single thread for serving.

torch defaults its intra-op thread pool to the host core count, ignoring process affinity, cgroup CPU quotas, and co-located worker processes. Any torch CPU op above the parallel_for grain size (32k elements) fans out across that pool, and the OpenMP workers then spin-wait after every parallel region - stealing cycles from the engine loop's serial code and, in containers, burning the CFS quota so that the entire process is throttled for the rest of the quota period. This is the root cause of the 2-6x structured-output decode throughput regression reported in #49013 (per-step pinning/fill of the ~12.6MB grammar bitmask on the hot path).

Previously only the multiproc executor mitigated this, by exporting `OMP_NUM_THREADS=1` before spawning workers - which also pinned model loading to a single thread, and never covered the uniproc (TP=1), Ray, or external-launcher paths. All #49013 reproductions are TP=1.

Manage torch threads explicitly in the workers instead:

- `Worker.init_device()`: set torch threads to available_cpus divided by the workers on this node, where available_cpus honors both scheduling affinity and the effective cgroup CPU quota (resolved from `/proc/self/cgroup`, v1 and v2). This keeps weight loading parallel but bounded - multiproc workers previously loaded single-threaded.
- After warmup completes in `compile_or_warm_up_model()`: set torch threads to 1. No steady-state CPU op benefits from intra-op parallelism; it only creates spin-wait contention and quota burn.
- The multiproc executor no longer exports `OMP_NUM_THREADS=1` to workers; the parent engine-core process still caps itself to 1 thread since the scheduler gains nothing from intra-op parallelism.

An externally-set `OMP_NUM_THREADS` is always respected; a value > 1 now logs a warning at the serving transition explaining the potential impact.

----

### Weight-loading time improvements

Qwen3-235B-A22B-Thinking-2507-FP8 (221 GiB), TP=4, enforce_eager=True, max_model_len=4096, gpu_memory_utilization=0.85, 4×GB200 / 144 CPU cores

**Warm cache** (4 back-to-back repeats per variant, ±1% spread):

| variant | weight load | total `LLM()` init | load as % of init |
|---|---|---|---|
| before | 37.52 s | 81.79 s | 45.9% |
| **after** | **31.60 s** | **76.06 s** | 41.5% |
| | **−15.8%** | **−7.0%** | |

**Cold cache** (~220 GiB evicted before each run; prefetch 14.2 s cold vs 0.8 s warm):

| variant | weight load | total init |
|---|---|---|
| before | 49.82 s | 95.88 s |
| **after** | **42.21 s** | **85.94 s** |
| | −15.3% | −10.4% |

----

Also [confirmed](https://github.com/vllm-project/vllm/pull/49168#issuecomment-5119258579) to have resolved the TP=1 stuctured output runtime performance regression:

Fixes #49013

### CODE REVIEW

**Comment by cjackal** [comment]
Hi, I think this change raises the following race condition in opening the shared objects:

```log
...
(Worker pid=1118) INFO 07-28 23:31:42 [torch_utils.py:234] Reducing Torch threads from 64 to 6 for startup.
Inconsistency detected by ld.so: ../elf/dl-tls.c: 613: _dl_allocate_tls_init: Assertion `listp != NULL' failed!
# repeated 8 times
...
```

The above error is raised for TP8 with 48 vcores (I'm using k8s so vcore), the change successfully reduces the threads to 6 (48 / 8) and hits the race condition on thread-local storage allocation.

**Comment by njhill** [comment]
Thanks @cjackal, I've reworked this in another commit that should hopefully fix this, though I wasn't able to repro it exactly. Would be grateful if you could validate it whenever you have the chance!

**Comment by cjackal** [comment]
> Thanks @cjackal, I've reworked this in another commit that should hopefully fix this, though I wasn't able to repro it exactly. Would be grateful if you could validate it whenever you have the chance!

Thank you for the prompt response!

I have tried the new commit and it seems the dl-tls regression is gone in the sense that the server launches successfully - but I was also told from my colleage that this type of error is often intermittent and should be tested in various environments, so I am not fully certain that there's no regression yet.

**Review by Isotr0py — APPROVED** [review]
Look reasonable!
